lib/luaxtable.c

```diff
@@ -0,0 +1,241 @@
+		struct xt_target *target;
+	} hook;
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static void luaxtable_release(void *private);
+static int luaxtable_reg_match(lua_State *L);
+static int luaxtable_reg_target(lua_State *L);
+
+/* Verifies the type of "opt" in the table at "idx" and returns true if the value can be extracted.
+ * If the value is not required, it returns false.
+ * If the value is required and the type is invalid, it raises an error.
+ *
+ * TODO: create a lunatik_checkfield() API to check the type of a table field and
+ * 	raise a proper error message
+ * */
+static bool luaxtable_reg_opt(lua_State *L, int idx, const char *opt, int type, bool req)
+{
+	lua_getfield(L, idx, opt);
+	int type_ = lua_type(L, -1);
+	if (type_ != type) {
+		// LUA_TNIL is 0 and LUA_TNONE is -1
+		if (type_ <= 0 && !req) {
+			/* the value was not required hence, should not try to extract it */
+			return false;
+		}
+		luaL_error(L, "invalid type for %s, expected %s", opt, lua_typename(L, type));
+	}
+
+	/* true if the value can be extracted (for optional parameters) */
+	return true;
+};
+
+#define LUAXTABLE_REG(mode, ttype)						\
+static int luaxtable_reg_##mode(lua_State *L) {				\
+	lunatik_object_t *object;					\
+	luaxtable_t *xtable_##mode;					\
+									\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	luaxtable_reg_opt(L, 1, "name", LUA_TSTRING, true);		\
+	const char *name = lua_tostring(L, -1);				\
+									\
+	luaxtable_reg_opt(L, 1, "revision", LUA_TNUMBER, true);		\
+	uint8_t revision = (uint8_t)lua_tointeger(L, -1);		\
+									\
+	luaxtable_reg_opt(L, 1, "family", LUA_TNUMBER, true);		\
+	unsigned short family = (unsigned short)lua_tointeger(L, -1);	\
+									\
+	unsigned proto = luaxtable_reg_opt(L, 1, "proto", LUA_TNUMBER, false) ? \
+			 (unsigned short)lua_tointeger(L, -1) : 0;	\
+									\
+	luaxtable_reg_opt(L, 1, "checkentry", LUA_TFUNCTION, true);	\
+	luaxtable_reg_opt(L, 1, "destroy", LUA_TFUNCTION, true);	\
+	luaxtable_reg_opt(L, 1, #mode, LUA_TFUNCTION, true);		\
+									\
+	struct xt_##mode *mode = lunatik_checkalloc(L, sizeof(struct xt_##mode));	\
+	memcpy((void *)mode->name, name, XT_EXTENSION_MAXNAMELEN);	\
+	mode->revision = revision;					\
+	mode->family = family;						\
+	mode->proto = proto;						\
+	mode->mode = luaxtable_##mode;					\
+	mode->checkentry = luaxtable_##mode##_check;			\
+	mode->destroy = luaxtable_##mode##_destroy;			\
+	mode->me = THIS_MODULE;						\
+									\
+	if (xt_register_##mode(mode) != 0) {				\
+		lunatik_free(mode);					\
+		luaL_error(L, "unable to register" #mode);		\
+	}								\
+									\
+	object = lunatik_newobject(L, &luaxtable_class, sizeof(luaxtable_t));	\
+	xtable_##mode = (luaxtable_t *)object->private;			\
+									\
+	xtable_##mode->runtime = lunatik_toruntime(L);			\
+	xtable_##mode->hook.mode = mode;				\
+	xtable_##mode->type = ttype;					\
+	lunatik_getobject(xtable_##mode->runtime);			\
+	lunatik_registerobject(L, 1, object);				\
```

Comment:
we can move this to an auxiliary function.. e.g., `luaxtable_newhook`

Reply:
This uses mode/hook, so we would again have a new macro or the auxiliary function will need to handle both cases under if..else or switch..case block